nearcore/src/config.rs

```diff
@@ -790,13 +790,58 @@ fn add_account_with_key(
     });
 }
 
-/// Generate a validator key and save it to the file path.
-fn generate_validator_key(account_id: AccountId, path: &Path) -> anyhow::Result<()> {
-    let signer = InMemoryValidatorSigner::from_random(account_id.clone(), KeyType::ED25519);
-    info!(target: "near", "Use key {} for {} to stake.", signer.public_key(), account_id);
-    signer
-        .write_to_file(path)
-        .with_context(|| format!("Error writing key file to {}", path.display()))
+enum KeyToGenerate<'a> {
+    Node,
+    Validator(Option<AccountId>),
+    TestValidator(AccountId, Option<&'a str>),
+}
+
+fn generate_key(
+    key: KeyToGenerate<'_>,
+    home_dir: &Path,
+    filename: &str,
+    overwrite: bool,
+) -> anyhow::Result<Option<InMemorySigner>> {
+    let path = home_dir.join(filename);
```

Comment:
Either works, but since joining happens in all places that call this function I figured it’s easier if it’s pushed to the function so it doesn’t have to be repeated over and over.